### PR TITLE
OCM-2641, OCM-17: Specify worker disk size for machine pools

### PR DESCRIPTION
**Pull request by JohnStrunk**
- [OCM-2641](https://issues.redhat.com//browse/OCM-2641): Adds `worker_disk_size` to cluster resource. This is used only at cluster creation to determine the size of the default MP's disk size
- [OCM-17](https://issues.redhat.com//browse/OCM-17): Adds `disk_size` to the machinepool resource. -- Only used at pool creation time

Supersceedes #100

### CODE REVIEW

**Comment by JohnStrunk** [comment]
@bardielle This is ready for a first look, though I want to do some more testing once the above bug is resolved.

File: provider/cluster_rosa_classic_resource.go [hunk 1]
```diff
@@ -459,6 +459,16 @@ func (t *ClusterRosaClassicResourceType) GetSchema(ctx context.Context) (result
 				},
 				Validators: adminCredsValidators(),
 			},
+			"worker_disk_size": {
+				Description: "Machine pool root disk size, in GiB.",
+				Type:        types.Int64Type,
+				Optional:    true,
+				Computed:    true,
+				PlanModifiers: []tfsdk.AttributePlanModifier{
+					tfsdk.UseStateForUnknown(),
```
Comment (bardielle) [inline provider/cluster_rosa_classic_resource.go:468]: I think we can add a validation here on the disk size value

File: internal/ocm/resource/cluster_test.go [hunk 2]
```diff
@@ -15,22 +15,22 @@ var _ = Describe("Cluster", func() {
 	})
 	Context("CreateNodes validation", func() {
 		It("Autoscaling disabled minReplicas set - failure", func() {
-			err := cluster.CreateNodes(false, nil, pointer(int64(2)), nil, nil, nil, nil, false)
+			err := cluster.CreateNodes(false, nil, pointer(int64(2)), nil, nil, nil, nil, false, nil)
```
Comment (nirarg) [inline internal/ocm/resource/cluster_test.go:18]: Please add one test with non empty diskSize value

File: internal/ocm/resource/cluster.go [hunk 3]
```diff
@@ -33,14 +33,22 @@ func (c *Cluster) Build() (object *cmv1.Cluster, err error) {
...
 	if computeMachineType != nil {
 		nodes.ComputeMachineType(
 			cmv1.NewMachineType().ID(*computeMachineType),
 		)
 	}
 
+	if workerDiskSize != nil {
+		nodes.ComputeRootVolume(
+			cmv1.NewRootVolume().AWS(
+				cmv1.NewAWSVolume().Size(int(*workerDiskSize)),
```
Comment (nirarg) [inline internal/ocm/resource/cluster.go:47]: Do we need to set default value here or just use the default value in the OCM backend?
What is the behavior in the CLI?

File: provider/machine_pool_resource.go [hunk 4]
```diff
@@ -184,6 +184,16 @@ func (t *MachinePoolResourceType) GetSchema(ctx context.Context) (result tfsdk.S
 					ValueCannotBeChangedModifier(),
 				},
 			},
+			"disk_size": {
```
Comment (nirarg) [inline provider/machine_pool_resource.go:187]: Please generate the attributes in the documentation

File: internal/ocm/resource/cluster.go (hunk 3 above)
Comment (leseb) [reply internal/ocm/resource/cluster.go:47]: In the CLI, if the size is not set or identical to the flavor's default size, we do nothing and let the backend handle the defaults.

File: provider/machine_pool_resource.go (hunk 4 above)
Comment (JohnStrunk) [reply provider/machine_pool_resource.go:187]: done

**Comment by nirarg** [comment]
/hold 
This feature is on hold
Will be merged only after v1.3.0 release

**Comment by JohnStrunk** [comment]
CI failure during cluster destroy:
```
[31m│[0m [0m[1m[31mError: [0m[0m[1mdeleting IAM Role (rhcsci-soph7-ControlPlane-Role): unable to detach instance profiles: AccessDenied: User: arn:aws:iam::425464789085:user/terraform-provider-minimal is not authorized to perform: iam:RemoveRoleFromInstanceProfile on resource: instance profile rhcsci-soph7-v9jl4-master-profile because no identity-based policy allows the iam:RemoveRoleFromInstanceProfile action
[31m│[0m [0m	status code: 403, request id: 33115eec-f165-4a5b-a4c7-3d908af11df0[0m
[31m│[0m [0m
[31m│[0m [0m[0m
[31m╵[0m[0m
[31m╷[0m[0m
[31m│[0m [0m[1m[31mError: [0m[0m[1mdeleting IAM Role (rhcsci-soph7-Worker-Role): unable to detach instance profiles: AccessDenied: User: arn:aws:iam::425464789085:user/terraform-provider-minimal is not authorized to perform: iam:RemoveRoleFromInstanceProfile on resource: instance profile rhcsci-soph7-v9jl4-worker-profile because no identity-based policy allows the iam:RemoveRoleFromInstanceProfile action
[31m│[0m [0m	status code: 403, request id: cd877cb3-7361-4d5d-9628-5afdeeca52bd[0m
```
Seems to be an AWS permission problem.

File: provider/machine_pool_resource.go (hunk 4 above)
Comment (bardielle) [reply provider/machine_pool_resource.go]: regenerate it please

File: provider/clusterrosaclassic/cluster_rosa_classic_resource.go [hunk 5]
```diff
@@ -252,6 +252,10 @@ func (r *ClusterRosaClassicResource) Schema(ctx context.Context, req resource.Sc
 					stringplanmodifier.RequiresReplace(),
 				},
 			},
+			"worker_disk_size": schema.Int64Attribute{
```
Comment (bardielle) [inline provider/clusterrosaclassic/cluster_rosa_classic_resource.go:255]: we should throw an error if the user tries to update it

**Comment by bardielle** [comment]
I had 2 comments but overall it looks good

File: provider/clusterrosaclassic/cluster_rosa_classic_resource.go (hunk 5 above)
Comment (JohnStrunk) [reply provider/clusterrosaclassic/cluster_rosa_classic_resource.go:255]: This option needs to be treated identically to what's being done in #228 since this setting is configuring the default MP.

Since the framework upgrade, we have removed all of the `ValueCannotBeChanged` checks. Should I look into re-introducing that?
And, what about #228-- should that be changed as well?
If so, I'd prefer to move these checks to another PR.

File: provider/machine_pool_resource.go (hunk 4 above)
Comment (JohnStrunk) [reply provider/machine_pool_resource.go]: done

File: provider/clusterrosaclassic/cluster_rosa_classic_resource.go (hunk 5 above)
Comment (bardielle) [reply provider/clusterrosaclassic/cluster_rosa_classic_resource.go:255]: I think it should also be changed in #228

**Comment by JohnStrunk** [comment]
Looks like CI troubles again:
```
2023-10-24T13:48:04.183Z [ERROR] provider.terraform-provider-aws_v5.22.0_x5: Response contains error diagnostic: @caller=github.com/hashicorp/terraform-plugin-go@v0.19.0/tfprotov5/internal/diag/diagnostics.go:58 tf_proto_version=5.4 tf_resource_type=aws_s3_bucket_policy tf_provider_addr=registry.terraform.io/hashicorp/aws @module=sdk.proto diagnostic_detail="" diagnostic_severity=ERROR tf_rpc=ApplyResourceChange diagnostic_summary="putting S3 Bucket (oidc-l1x6) Policy: operation error S3: PutBucketPolicy, https response error StatusCode: 403, RequestID: XVKX5BJF6YSHM3FF, HostID: m20I89LXf7YOjhYoGQoJpiVBtKn9wZMY0PyljlpKjMkTklb300pcS/J5j2mfCsPjrXnpTKwSwxdiTKp9vVz/ag==, api error AccessDenied: Access Denied" tf_req_id=b88605aa-841c-fb26-e157-5ed8fce16572 timestamp=2023-10-24T13:48:04.182Z
2023-10-24T13:48:04.185Z [ERROR] vertex "module.oidc_config_input_resources.module.rosa_oidc_config_resources[0].aws_s3_bucket_policy.allow_access_from_another_account" error: putting S3 Bucket (oidc-l1x6) Policy: operation error S3: PutBucketPolicy, https response error StatusCode: 403, RequestID: XVKX5BJF6YSHM3FF, HostID: m20I89LXf7YOjhYoGQoJpiVBtKn9wZMY0PyljlpKjMkTklb300pcS/J5j2mfCsPjrXnpTKwSwxdiTKp9vVz/ag==, api error AccessDenied: Access Denied
```

```
putting S3 Bucket (oidc-l1x6) Policy: operation error S3: PutBucketPolicy, https response error StatusCode: 403, RequestID: XVKX5BJF6YSHM3FF, HostID: m20I89LXf7YOjhYoGQoJpiVBtKn9wZMY0PyljlpKjMkTklb300pcS/J5j2mfCsPjrXnpTKwSwxdiTKp9vVz/ag==, api error AccessDenied: Access Denied

with module.oidc_config_input_resources.module.rosa_oidc_config_resources[0].aws_s3_bucket_policy.allow_access_from_another_account,
 on .terraform/modules/oidc_config_input_resources/oidc_config_resources/create_s3_bucket.tf line 21, in resource "aws_s3_bucket_policy" "allow_access_from_another_account":
  21: resource "aws_s3_bucket_policy" "allow_access_from_another_account"
```
